ipv8/gossip/community.py

```diff
@@ -0,0 +1,504 @@
+                if payload not in bloomfilter:
+                    self.send_to_key(message.public_key, GossipRule.DEFAULT, message.target_public_key, payload)
+
+    def _pack_gossip_message(self, rule, target_public_key, message):
+        """
+        Pack a gossip message
+
+        :param rule: the rule passed in the message
+        :param target_public_key: the public key of the target peer
+        :param message: the contents of the message
+        :return: the packed gossip message
+        """
+        return self.prefix + self.serializer.pack_multiple(GossipMessage(self.my_peer.key, rule, target_public_key,
+                                                                         message).to_pack_list())
+
+    def _unpack_gossip_message(self, data):
+        """
+        Unpack a gossip message
+
+        :param data: the raw message
+        :return: the unpacked message
+        """
+        message = self.serializer.unpack_to_serializables([GossipMessage, ], data[self.PREFIX_LENGTH:])[0]
+        message.rule = int(message.rule)
+
+        return message
+
+    def _vote_suppress(self, message):
+        """
+        Attempt to suppress a give peer, as specified in the message. This method should only be called when the
+        message source is indeed a trusted neighbor.
+
+        :param message: the received SUPPRESS message
+        :return: None
+        """
+        self._vote_change_rule(message, GossipRule.SUPPRESS, self._suppress_peer)
+
+    def _suppress_peer(self, target_public_key):
+        """
+        Suppress a peer identified by its public key
+
+        :param target_public_key: the public key of the peer which should be suppressed
+        :return: None
+        """
+        self.purge(target_public_key)
+        self.rules_db[target_public_key] = GossipRule.SUPPRESS
+
+    def _vote_change_rule(self, message, rule, action):
+        """
+        Attempt to change the gossip rule for a give peer, as specified in the message. This method should only
+        be called when the message source is indeed a trusted neighbor.
+
+        :param message: the received SUPPRESS message
+        :param rule: the rule to which one should change the peer indicated in the message
+        :param action: a function which should be called when the vote passes. It should have the following signature:
+                       (target_public_key : str) => None
+        :return: None
+        """
+        assert message.target_public_key not in self.rule_change_db or \
+               isinstance(self.rule_change_db[message.target_public_key], GossipRuleChangeBallot), \
+            "The contents of the rule_change_db for this target_public_key are not correct."
+
+        if self.rule_change_db.get(message.target_public_key, None) is None:
+            callback_id = reactor.callLater(VOTE_TIMEOUT, self._reset_vote, message.target_public_key)
```

Comment:
Please use a deferLater which is properly `register_task`'d.

Reply:
Most of the examples of the `register_task` calls, which I've seen in the codebase were being used together with a `reactor.callLater`, that's why I opted for this. I'll change it to `deferLater`, but it will require some extra modifications, I think.

Reply:
You can also register task this indeed. It's just that `Deferred`s are easier to handle than `DelayedCall`s.